### PR TITLE
Create the notion of unstable extensions

### PR DESCRIPTION
**Pull request by tristan957**
As a DBaaS provider, Neon needs to provide a stable platform for customers to build applications upon. At the same time however, we also need to enable customers to use the latest and greatest technology, so they can prototype their work, and we can solicit feedback. If all extensions are treated the same in terms of stability, it is hard to meet that goal.

There are now two new GUCs created by the Neon extension:

neon.allow_unstable_extensions: This is a session GUC which allows a session to install and load unstable extensions.

neon.unstable_extensions: This is a comma-separated list of extension names. We can check if a CREATE EXTENSION statement is attempting to install an unstable extension, and if so, deny the request if neon.allow_unstable_extensions is not set to true.

### CODE REVIEW

File: pgxn/neon/unstable_extensions.c
Comment (ololobus) [inline pgxn/neon/unstable_extensions.c:105]: More of a question: I'm not sure we need to do anything with alter. If you already installed it, you kinda gave a consent that it's unstable

File: test_runner/regress/test_unstable_extensions.py
Comment (ololobus) [inline test_runner/regress/test_unstable_extensions.py:39]: You test only restriction path, but how you gonna install `pg_prewarm`? THat's a follow up to the GUC definition.

The test needs a `neon.allow_unstable_extensions` exercise path. And if it's USERSET, you could just set it per session or per transaction

File: pgxn/neon/unstable_extensions.c
Comment (ololobus) [inline pgxn/neon/unstable_extensions.c:18]: How is this supposed to work for versions < 16 if number of args is different? And I see that tests with these versions failed

It probably makes sense to just use the standard Postgres' pfree, palloc, pstrdup

File: pgxn/neon/unstable_extensions.c
Comment (ololobus) [inline pgxn/neon/unstable_extensions.c:52]: cc @knizhnik to review the C strings trickery :) Not sure I can do it quickly today

File: pgxn/neon/unstable_extensions.c
Comment (tristan957) [reply pgxn/neon/unstable_extensions.c]: See pgversioncompat changes

File: pgxn/neon/unstable_extensions.c
Comment (ololobus) [inline pgxn/neon/unstable_extensions.c:149]: Real users don't have superuser, I think this should be PGC_USERSET. Otherwise it just won't work

You probably misread my comments, I suggested PGC_SUSET for `neon.unstable_extensions` https://github.com/neondatabase/neon/pull/9508/files#r1816479235

File: pgxn/neon/unstable_extensions.c
Comment (tristan957) [reply pgxn/neon/unstable_extensions.c:49]: There is no trickery, just usual flexible array type stuff at the end of the struct fwiw.

**Review by ololobus — APPROVED** [review]
LGTM

**Comment by tristan957** [comment]
Is this good to merge, or should we wait for another review?

File: pgxn/neon/unstable_extensions.c
Comment (ololobus) [reply pgxn/neon/unstable_extensions.c:158]: > Why unstable_extensions_str is can be set only by superuser (PGC_SUSET), while disabling this check can be done by normal user (PGC_USERSET)? It seems to be some contradiction here...

No, the point is that only us can specify the list of unstable extensions, but we want users to be able to set the flag to confirm that they really want to install them

**Comment by tristan957** [comment]
Superseded by https://github.com/neondatabase/neon/pull/9540.

**Pull request closed without merging**
